File: cpp/luhn/luhn.cpp
#include "luhn.h"
#include <algorithm>
#include <cctype>
#include <iostream>

namespace luhn
{
    bool valid(std::string number)
    {

        number.erase(std::remove_if(number.begin(), number.end(), ::isspace), number.end());
        if (number.length() <= 1) return false;

        int total{0};
        int digit{0};
        for (int i = number.length() - 1; i >= 0; i--)
        {
            if (!isdigit(number[i])) return false;

            int value = number[i] - '0';
            if ((digit + 1) % 2 == 0)
            {
                (value * 2 > 9) ? total += value * 2 - 9 : total += value * 2;
                digit++;
                continue;
            }

            total += value;
            digit++;
        }

        return total % 10 == 0;
    }
} // namespace luhn

File: go/luhn/luhn.go
package luhn

import (
	"strings"
	"unicode"
)

func Valid(id string) bool {
	id = strings.ReplaceAll(id, " ", "")
	if len(id) <= 1 {
		return false
	}

	sum := 0
	count := 0
	for i := len(id) - 1; i >= 0; i-- {
		if !unicode.IsDigit(rune(id[i])) {
			return false
		}

		digit := int(id[i] - '0')
		if count%2 != 0 {
			doubled := digit * 2
			if doubled > 9 {
				doubled -= 9
			}
			sum += doubled
			count++
			continue
		}
		sum += digit
		count++
	}

	return sum%10 == 0
}
